File: lintian_brush/dummy_transitional.py
#!/usr/bin/python3
# Copyright (C) 2022 Jelmer Vernooij
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA

import argparse
import asyncio
from dataclasses import dataclass
import logging
import re
import sys
from typing import Dict, Set
import warnings
import yaml

from debian.deb822 import PkgRelation
from debmutate.control import suppress_substvar_warnings


def parse_relations(t: str):
    with warnings.catch_warnings():
        suppress_substvar_warnings()
        return PkgRelation.parse_relations(t)


REGEXES = [
    r'.*\((.*, )?(dummy )?transitional (dummy )?package\)',
    r'.*\((.*, )?(dummy )?transitional (dummy )?package for ([^ ]+)\)',
    r'.*\(transitional development files\)',
    r'.*\(transitional\)',
    r'.* [-—] transitional( package)?',
    r'.*\[transitional package\]',
    r'.* - transitional (dummy )?package',
    r'transitional package -- safe to remove',
    r'(dummy )?transitional (dummy )?package (for|to) (.*)',
    r'transitional dummy package',
    r'transitional dummy package: ([^ ]+)',
    r'transitional package, ([^ ]+)',
    r'(dummy )?transitional (dummy )?package, ([^ ]+) to ([^ ]+)',
    r'transitional package( [^ ]+)?',
    r'([^ ]+) transitional package',
    r'.* transitional package',
    r'.*transitional package for .*',
]


@dataclass
class TransitionalPackage:

    from_name: str
    to_expr: str

    def json(self):
        return {'from_name': self.from_name, 'to_expr': self.to_expr}


async def find_reverse_dependencies(udd, package):
    by_source: Dict[str, Set[str]] = {}
    fields = ['recommends', 'depends', 'pre_depends', 'enhances',
              'suggests', 'provides']
    query = (
        "SELECT source, package, " + ", ".join(fields) +
        " FROM packages WHERE " + " OR ".join([
            f"{field} LIKE CONCAT('%', $1::text, '%')" for field in fields]))
    for row in await udd.fetch(query, package):
        for relname in fields:
            parsed = parse_relations(row[relname] or '')
            for option in parsed:
                for rel in option:
                    if rel['name'] == package:
                        by_source.setdefault(
                            row['source'], set()).add(row['package'])

    fields = ["build_depends", "build_depends_indep", "build_depends_arch",
              "build_conflicts", "build_conflicts_indep",
              "build_conflicts_arch"]
    query = ("SELECT source, " + ", ".join(fields) +
             " FROM sources WHERE " + " OR ".join([
                 f"{field} LIKE CONCAT('%', $1::text, '%')"
                 for field in fields]))
    for row in await udd.fetch(query, package):
        for relname in fields:
            parsed = parse_relations(row[relname] or '')
            for option in parsed:
                for rel in option:
                    if rel['name'] == package:
                        by_source.setdefault(row['source'], set())
    return by_source


async def find_dummy_transitional_packages(udd, release):
    ret = {}
    query = """\
SELECT package, description, depends FROM packages
WHERE release = $1 AND description LIKE '%transitional%'"""
    for row in await udd.fetch(query, release):
        for regex in REGEXES:
            if re.fullmatch(regex, row[1]):
                break
        else:
            logging.debug(
                'Unknown syntax for dummy package description: %r', row[1])
        if row[2] is None:
            logging.debug('no replacement for %s', row[0])
        else:
            depends = parse_relations(row[2])
            if len(depends) != 1:
                logging.debug(
                    'no single transition target for %s: %r', row[0], row[2])
        ret[row[0]] = TransitionalPackage(
            from_name=row[0], to_expr=row[2])  # type: ignore
    return ret


async def main(argv=None):
    from .udd import connect_udd_mirror
    parser = argparse.ArgumentParser(prog="candidates")
    parser.add_argument("--release", type=str, default="sid")
    parser.add_argument('--list-transitional-dummy', action="store_true")
    parser.add_argument('--list-uses-transitional-dummy', action="store_true")
    args = parser.parse_args()

    logging.basicConfig(level=logging.INFO, format='%(message)s')

    async with await connect_udd_mirror() as udd:
        transitions = await find_dummy_transitional_packages(udd, args.release)

        if args.list_transitional_dummy:
            yaml.dump(
                [transition.json() for transition in transitions.values()],
                sys.stdout)
        elif args.list_uses_transitional_dummy:
            for dep in transitions:
                by_source = await find_reverse_dependencies(udd, dep)
                for source, binaries in by_source.items():
                    for binary in binaries:
                        print(f'{source} / {binary} / {dep}')
        else:
            parser.print_usage()
            return 1


if __name__ == '__main__':
    sys.exit(asyncio.run(main(sys.argv[1:])))

File: scrub-obsolete/src/dummy_transitional.rs
use debian_control::lossless::relations::{Entry, Relations};
use serde::Serialize;
use sqlx::{PgPool, Row};
use std::collections::{HashMap, HashSet};

lazy_static::lazy_static! {
    pub static ref REGEXES: Vec<regex::Regex> = vec![
        regex::Regex::new(r".*\((.*, )?(dummy )?transitional (dummy )?package\)").unwrap(),
        regex::Regex::new(r".*\((.*, )?(dummy )?transitional (dummy )?package for ([^ ]+)\)").unwrap(),
        regex::Regex::new(r".*\(transitional development files\)").unwrap(),
        regex::Regex::new(r".*\(transitional\)").unwrap(),
        regex::Regex::new(r".* [-—] transitional( package)?").unwrap(),
        regex::Regex::new(r".*\[transitional package\]").unwrap(),
        regex::Regex::new(r".* - transitional (dummy )?package").unwrap(),
        regex::Regex::new(r"transitional package -- safe to remove").unwrap(),
        regex::Regex::new(r"(dummy )?transitional (dummy )?package (for|to) (.*)").unwrap(),
        regex::Regex::new(r"transitional dummy package").unwrap(),
        regex::Regex::new(r"transitional dummy package: ([^ ]+)").unwrap(),
        regex::Regex::new(r"transitional package, ([^ ]+)").unwrap(),
        regex::Regex::new(r"(dummy )?transitional (dummy )?package, ([^ ]+) to ([^ ]+)").unwrap(),
        regex::Regex::new(r"transitional package( [^ ]+)?").unwrap(),
        regex::Regex::new(r"([^ ]+) transitional package").unwrap(),
        regex::Regex::new(r".* transitional package").unwrap(),
        regex::Regex::new(r".*transitional package for .*").unwrap(),
    ];
}

#[derive(Debug)]
pub struct TransitionalPackage {
    pub from_name: String,
    pub to_expr: String,
}

impl Serialize for TransitionalPackage {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: serde::Serializer,
    {
        use serde::ser::SerializeMap;
        let mut map = serializer.serialize_map(Some(2))?;
        map.serialize_entry("from_name", &self.from_name)?;
        map.serialize_entry("to_expr", &self.to_expr)?;
        map.end()
    }
}

pub async fn find_reverse_dependencies(
    udd: &PgPool,
    package: &str,
) -> Result<HashMap<String, HashSet<String>>, sqlx::Error> {
    let mut by_source = HashMap::new();
    let fields = &[
        "recommends",
        "depends",
        "pre_depends",
        "enhances",
        "suggests",
        "provides",
    ];

    let mut builder = sqlx::QueryBuilder::new("SELECT source, package, ");

    for (i, field) in fields.iter().enumerate() {
        if i > 0 {
            builder.push(", ");
        }
        builder.push(field);
    }

    builder.push(" FROM packages WHERE ");

    for (i, field) in fields.iter().enumerate() {
        if i > 0 {
            builder.push(" OR ");
        }

        builder.push(&format!("{} LIKE CONCAT('%', ", field));
        builder.push_bind(package);
        builder.push("::text, '%%')");
    }

    let query = builder.build();

    for row in query.fetch_all(udd).await? {
        let source: String = row.get("source");
        let binary: String = row.get("package");
        for field in fields {
            let value: String = row.get(field);
            let parsed: Relations = value.parse().unwrap();
            for entry in parsed.entries() {
                for rel in entry.relations() {
                    if rel.name() == package {
                        by_source
                            .entry(source.clone())
                            .or_insert_with(HashSet::new)
                            .insert(binary.clone());
                    }
                }
            }
        }
    }

    let fields = &[
        "build_depends",
        "build_depends_indep",
        "build_depends_arch",
        "build_conflicts",
        "build_conflicts_indep",
        "build_conflicts_arch",
    ];

    let mut builder = sqlx::QueryBuilder::new("SELECT source, ");
    for (i, field) in fields.iter().enumerate() {
        if i > 0 {
            builder.push(", ");
        }
        builder.push(field);
    }
    builder.push(" FROM sources WHERE ");
    for (i, field) in fields.iter().enumerate() {
        if i > 0 {
            builder.push(" OR ");
        }
        builder.push(&format!("{} LIKE CONCAT('%', ", field));
        builder.push_bind(package);
        builder.push("::text, '%%')");
    }
    let query = builder.build();

    for row in query.fetch_all(udd).await? {
        let source: String = row.get("source");
        for field in fields {
            let value: String = row.get(field);
            let parsed: Relations = value.parse().unwrap();
            for option in parsed.entries() {
                for rel in option.relations() {
                    if rel.name() == package {
                        by_source.entry(source.clone()).or_insert_with(HashSet::new);
                    }
                }
            }
        }
    }
    Ok(by_source)
}

pub async fn find_dummy_transitional_packages(
    udd: &PgPool,
    release: &str,
) -> Result<HashMap<String, TransitionalPackage>, sqlx::Error> {
    let mut ret = HashMap::new();

    let rows = sqlx::query_as::<_, (String, String, Option<String>)>(
        r#"
        SELECT package, description, depends
        FROM packages
        WHERE release = $1 AND description LIKE '%transitional%'
        "#,
    )
    .bind(release)
    .fetch_all(udd)
    .await?;

    for row in rows {
        let r = if let Some(regex) = REGEXES.iter().find(|regex| regex.is_match(&row.1)) {
            regex
        } else {
            log::debug!("Unknown syntax for dummy package description: {:?}", row.1);
            continue;
        };
        log::debug!("{}: {:?}", row.0, r);
        if let Some(depends) = row.2 {
            let depends: Relations = depends.parse().unwrap();
            let mut entries = depends.entries();
            let e = if let Some(e) = entries.next() {
                e
            } else {
                Entry::new()
            };
            if entries.next().is_some() {
                log::debug!("no single transition target for {}: {:?}", row.0, depends);
                continue;
            }
            ret.insert(
                row.0.clone(),
                TransitionalPackage {
                    from_name: row.0,
                    to_expr: e.to_string(),
                },
            );
        } else {
            log::debug!("no replacement for {}", row.0);
        }
    }
    Ok(ret)
}
